src/decorated_surface.rs

```diff
@@ -329,23 +345,42 @@ impl<H: Handler> DecoratedSurface<H> {
     /// This string may be used to identify the surface in a task bar, window list, or other user
     /// interface elements provided by the compositor.
     pub fn set_title(&self, title: String) {
-        self.shell_surface.set_title(title);
+        match self.shell_surface {
+            shell::Surface::Xdg(ref xdg) => {
+                xdg.toplevel.set_title(title);
+            },
+            shell::Surface::Wl(ref wl) => {
+                wl.set_title(title);
+            },
+        }
     }
 
     /// Set a class for the surface.
     ///
     /// The surface class identifies the general class of applications to which the surface
     /// belongs. A common convention is to use the file name (or the full path if it is a
     /// non-standard location) of the application's .desktop file as the class.
+    ///
+    /// When using xdg-shell, this calls `ZxdgTopLevelV6::set_app_id`.
+    /// When using wl-shell, this calls `WlShellSurface::set_class`.
     pub fn set_class(&self, class: String) {
-        self.shell_surface.set_class(class);
+        match self.shell_surface {
+            shell::Surface::Xdg(ref xdg) => {
+                xdg.toplevel.set_app_id(class);
+            },
+            shell::Surface::Wl(ref wl) => {
+                wl.set_class(class);
+            },
+        }
     }
 
     /// Turn on or off decoration of this surface
     ///
     /// Automatically disables fullscreen mode if it was set.
     pub fn set_decorate(&mut self, decorate: bool) {
-        self.shell_surface.set_toplevel();
+        if let shell::Surface::Wl(ref surface) = self.shell_surface {
+            surface.set_toplevel();
+        }
```

Comment:
We need to unset fullscreen mode on xdg_shell too, by calling `XdgToplevelV6::unset_fullscreen()`